### PR TITLE
base class exception default arguments

### PR DESCRIPTION
**Pull request by frosty00**
lmk ur thoughts

### CODE REVIEW

**Comment by kroitor** [comment]
@frosty00 looking great! Thx so much!

Now I'm thinking of making it forward-compatible so that we could extend it in the future and maybe add one or more fields to it if/when we want.

What if we edit the constructor slightly from this (pseudocode):

```JavaScript
// `code` and `previous` are listed in php only
constructor (message, exchangeId, foo=undefined, bar=undefined, code=0, previous=undefined) {
   this.exchangeId = exchangeId
   this.foo = foo
   this.bar = bar
   ...
}
```

to this (also pseudocode):

```JavaScript
// `code` and `previous` are listed in php only
constructor (message, params={}, code=0, previous=undefined) {
    // deep-extend the entire exception object itself just like an `Exchange` instance does in the constructor
    this.assign (deepExtend (this, params)) // assign all properties from params={}, whatever they are
}
```

↑ Does that make sense to you? That actually helps achieve several goals:

- we add the capability of extending it (if we decide to add another property later)
- it still fully conforms to the native built-in base exception constructor (the `message` is still a separate first-argument)
- it will remain portable when we add more languages

I only see benefits in wrapping the default args in a `params={}` dict, however, can't see a drawback. 

Let me know if you agree (I may be missing something).

**Comment by frosty00** [comment]
> I only see benefits in wrapping the default args in a params={} dict, and I don't see any downsides.

I would rather keep it using defined arguments, since the former is too duck typeish.

```

    we add the capability of extending it (if we decide to add another property later)
    it still fully conforms to the native built-in base exception constructor (the message is still a separate first-argument)
    it will remain portable when we add more languages
```

everything that can be done your way can also be done in most languages the first way, and it is clear what properties errors support for the user to read. mind you, i still have a toothache...

**Comment by kroitor** [comment]
@frosty00 in Python you can specify certain arguments by their name, without having to specify them all, while in JS or PHP you will have to list all the undefined args before you reach a defined arg. Suppose you then want to add a new arg at some point later, that is not related to HTTP.

Does that mean that if we add a non-http exception later we would throw it in JS like so:

```
throw new NonHttpError (message, undefined, undefined, undefined, ..., undefined, nonHttpArg)
```

? )

**Comment by frosty00** [comment]
my idea was that we were going to set all of these error properties from handleErrors, where they are already all defined. 

what do you think about moving the staticmethods to a new file in python, i can revert that if you'd like? however there may be some advantages to my layout in the future to avoid circular deps

**Comment by kroitor** [comment]
> my idea was that we were going to set all of these error properties from handleErrors, where they are already all defined.

We sometimes raise errors from places outside of handleErrors – `ArgumentsRequired` is one such example. If you do a quick search for `throw` in the js/* – you'll find them all.

> what do you think about moving the staticmethods to a new file in python, i can revert that if you'd like? 

At this stage with only two methods it adds a layer of indirection that is somewhat less convenient that it was – you read the method name in one file and then you have to read the body in another file. However, if this is going to be extended, then it might introduce some convenience – for now I don't really see how splitting that code and keeping the staticmethod in place can help us further.

> however there may be some advantages to my layout in the future to avoid circular deps

It will be great if you elaborate a bit more on on that )

**Comment by frosty00** [comment]
> We sometimes raise errors from places outside of handleErrors – ArgumentsRequired is one such example. If you do a quick search for throw in the js/* – you'll find them all.

right, however the only unified error properties something that is not raised in `handleErrors` needs to have is `exchangeId`

> for now I don't really see how splitting that code and keeping the staticmethod in place can help us further.

I agree, i think i got carried away, and it doesn't really help us tbh

File: js/base/functions/string.js [hunk 1]
```diff
@@ -11,10 +11,9 @@ module.exports =
 
       // hasFetchOHLCV → has_fetch_ohlcv; parseHTTPResponse → parse_http_response
     , unCamelCase: s => s.match (/^[A-Z0-9_]+$/) ? s : (s.replace (/[a-z0-9][A-Z]/g, x => x[0] + '_' + x[1]).replace(/[A-Z0-9][A-Z0-9][a-z]/g, x => x[0] + '_' + x[1] + x[2]).toLowerCase ())
+    , camelCase: s => s.replace (/_(\w)/g, (x, y) => y.toUpperCase ())
```
Comment (kroitor) [inline js/base/functions/string.js]:
```
> 'fetch_ohlcv'.replace (/_(\w)/g, (x, y) => y.toUpperCase ())
'fetchOhlcv'
```

Comment (frosty00) [reply js/base/functions/string.js]: I'm just using that method to convert between camelCase properties of errors and underscore case. Will change to storing the properties in the underscore case and aliasing the camelCase properties

**Comment by frosty00** [comment]
this can be merged, give it a test. basically in all languages

```
error = ccxt.ExchangeError ('message', 'kucoin')

error.exchangeId === error.exchange_id

error.exchangeId = 'something else'

console.log (error.exchange_id)
```

^ this should work

**Comment by hroff-1902** [comment]
C'mon, guys, we need it...

File: js/base/errors.js [hunk 2]
```diff
@@ -17,19 +18,34 @@ function subclass (BaseClass, classes, namespace = {}) {
 
             [className]: class extends BaseClass {
 
-                constructor (message) {
+                constructor (errorMessage, verbose = false, exchangeId = undefined, httpStatusCode = undefined, httpStatusText = undefined, url = undefined, httpMethod = undefined, responseHeaders = undefined, responseBody = undefined, responseJson = undefined) {
```
Comment (kroitor) [inline js/base/errors.js]: Suppose we need to add `exchangeMethod`, `methodArguments`, `requestHeaders` and `requestBody`, what would you prefer:

```
constructor (
    errorMessage,
    verbose = false,
    exchangeId = undefined,
    exchangeMethod = undefined,
    exchangeMethodArguments = undefined,
    httpStatusCode = undefined,
    httpStatusText = undefined,
    url = undefined,
    httpMethod = undefined,
    requestHeaders = undefined,
    requestBody = undefined,
    responseHeaders = undefined,
    responseBody = undefined,
    responseJson = undefined
)
```

or

```
constructor (
    errorMessage,
    verbose = false,
    params = {
        exchangeId = undefined,
        exchangeMethod = undefined,
        exchangeMethodArguments = undefined,
        httpStatusCode = undefined,
        httpStatusText = undefined,
        url = undefined,
        httpMethod = undefined,
        requestHeaders = undefined,
        requestBody = undefined,
        responseHeaders = undefined,
        responseBody = undefined,
        responseJson = undefined
    }
)
```

? )

**Pull request closed without merging**